### PR TITLE
Lightning: Support for payment comments, Tor mint support

### LINKED ISSUE #48: [Feature] Support entering comment when paying to LNURL or Lightning address
There is currently no option to enter the comment.

### PR DESCRIPTION
**Pull request by KraXen72**
closes #48, closes #75

## lightning: payment comment support
**EDIT:** this is out of date. comment is now edited during amount editing stage. invoice is only encoded once.
The comment has to be encoded in the invoice for it to be sent.  
Since the whole system is somewhat complicated, and i couldn't make sense of it, i made this small diagram.  
Stuff in green is what i changed.  
image
  
We have two options 
- re-encode the invoice whenever the user hits done on the comment input field:
```diff
- extra loading time whenever you hit done. kind of bad UX.
+ no need to encode the invoice when hitting the `Pay now` button
```
- only encode the comment when user hits `Pay now`
```diff
+ no extra loading time when changing the comment, can do many edits to it quickly
- extra loadig time when hitting `Pay now`
```
## Tor mint support
- [x] added better error message when Tor VPN is not running

### CODE REVIEW

**Comment by KraXen72** [comment]
thanks for the pointing comments, i'll follow them. since this pr is marked as a draft, it's not done yet.

**Comment by KraXen72** [comment]
@minibits-cash i've tested it and it worked, however, there are currently a few issues with it:
- whenever i hit the `done` button on the memo input field, i call `onAmountInputEndEditing` to re-calculate the encoded invoice url, with the new comment url parameter. this creates a bit of (unnecessary?) lag/waiting time for the user. can we do something about this? 
    - maybe the invoice url could just be re-calculated once, with the comment, when hitting pay, instead of whenever the comment is edited (not entirely sure how i'd do this though)
- whenever i hit done on the comment editing, whatever mint was selected as the one to pay from gets reset to the first one in the list. I've tried fixing this but to no avail.
- You can't "submit" or save the comment without hitting done (the input field will still be focused even when you hide the keyboard or tap elsewhere)

I was also thinking about maybe moving the comment input to the first stage, when the amount is inputted, but i'm not sure about it.
please let me know your opinion these problems. Thanks!

File: src/services/lnurlService.ts [hunk 1]
```diff
@@ -123,27 +123,27 @@ const getLnurlAddressParams = async (lnurlAddress: string) => {
...
-const getInvoice = async(lnurlParams: LNURLPayParams, amountMsat?: number) => {
+const getInvoice = async (lnurlParams: LNURLPayParams, amountMsat?: number, comment?: string) => {
+    if (!amountMsat || amountMsat === 0) { amountMsat = lnurlParams.minSendable }
 
-    if(!amountMsat || amountMsat === 0) {
-        amountMsat = lnurlParams.minSendable
-    }
-
-    const url = lnurlParams.callback.includes('?') ? `${lnurlParams.callback}&amount=${amountMsat}` : `${lnurlParams.callback}?amount=${amountMsat}`
+    const url = new URL(lnurlParams.callback)
```
Comment (minibits-cash) [inline src/services/lnurlService.ts:129]: 👍

Comment (KraXen72) [reply src/services/lnurlService.ts:129]: yeah i was looking at it and though, wait a minute, this can be done nicely with `URLSearchParams` :)

**Comment by KraXen72** [comment]
you might want to load the large diffs as well - i avoided re-formatting as best i could, but maybe prettier's `format selection` command had a few side-effects and formatted more than i was currently working with...

**Comment by KraXen72** [comment]
As a part of this pull request i will also include some small changes to the MintInfoScreen to fix bugs reported by a user on Telegram

**Comment by KraXen72** [comment]
@minibits-cash i have updated the PR description and https://github.com/minibits-cash/minibits_wallet/pull/83#issuecomment-2184131272 comment, please see the changes i've made.
I've tested it by sending sats to Wallet of Satoshi and it works.

**Comment by KraXen72** [comment]
there is one more bug that sometimes happens:
`'Not ready... exiting',` when i enter the comment and hit confirm.
it causees the comment and amount to reset, and when you input it again, it works. 
it doesen't happen all the time, so i can't reliably reproduce it yet to fix it.

**Comment by KraXen72** [comment]
please do not merge yet, i'm planning to add the Tor guide.

**Comment by minibits-cash** [comment]
> there is one more bug that sometimes happens: `'Not ready... exiting',` when i enter the comment and hit confirm. it causees the comment and amount to reset, and when you input it again, it works. it doesen't happen all the time, so i can't reliably reproduce it yet to fix it.
> 
> it really mostly doesen't happen.

Could not simulate so far.

**Comment by minibits-cash** [comment]
Recent changes - Fixed that lnurl comment did not make it to transaction detail of the payer. Quick fix that payment preimage was not logged into the audit trail. Made comment visible but consuming less space after invoice is generated.

Tested that comment is correctly passed to Wallet of Satoshi and Minibits wallet of the payee.

Still todo: test non-lnurl transfer types to avoid regressions.

**Comment by KraXen72** [comment]
nice improvements! thanks!

**Comment by KraXen72** [comment]
i tested an ecash payment from minibits to minibits (official mint) with and without a memo and both worked well without any issues.

**Comment by minibits-cash** [comment]
> i tested an ecash payment from minibits to minibits (official mint) with and without a memo and both worked well without any issues.

I meant to test non-lnurl based transfers, donations, payment requests processed in TransferScreen, not ecash sends that we did not touch.

**Comment by minibits-cash** [comment]
> > I meant to test non-lnurl based transfers, donations, payment requests processed in TransferScreen, not ecash sends that we did not touch.
> 
> @minibits-cash how can i test these? thanks.

Basic flow:
1. Create invoice via Receive/Topup or in any Lightning or ecash wallet
2. Pay it from Minibits by scanning / pasting the invoice

Payment request (in Minibits this is lightning invoice shared with a contact over Nostr relay)
1. Receive payment request from anothe Minibits wallet
2. Pay it from Minibits

Donation:
1. Book custom wallet address
2. Pay generated invoice from the same Minibits wallet

**Review by minibits-cash — APPROVED** [review]

**Pull request merged**